server/plugins/LFUGens.cpp

```diff
@@ -3197,56 +3200,24 @@ void IEnvGen_Ctor(IEnvGen* unit) {
         SETCALC(IEnvGen_next_k);
     }
 
-    // pointer, offset
-    // initlevel, numstages, totaldur,
-    // [dur, shape, curve, level] * numvals
     int numStages = (int)IN0(3);
     int numvals = numStages * 4; // initlevel + (levels, dur, shape, curves) * stages
-    float offset = unit->m_offset = IN0(1);
-    float point = unit->m_pointin = IN0(0) - offset;
     unit->m_envvals = (float*)RTAlloc(unit->mWorld, (int)(numvals + 1.) * sizeof(float));
     ClearUnitIfMemFailed(unit->m_envvals);
 
+    // fill m_envvals with the values
     unit->m_envvals[0] = IN0(2);
-    //  Print("offset of and initial  values %3,3f, %3.3f\n", offset, unit->m_envvals[0]);
-    // fill m_envvals with the values;
     for (int i = 1; i <= numvals; i++) {
         unit->m_envvals[i] = IN0(4 + i);
-        //      Print("val for: %d, %3.3f\n", i, unit->m_envvals[i]);
     }
 
-    //  float out = OUT0(0);
-    float totalDur = IN0(4);
-    float level = 0.f;
-    float newtime = 0.f;
-    int stage = 0;
-    float seglen = 0.f;
-    if (point >= totalDur) {
-        unit->m_level = level = unit->m_envvals[numStages * 4]; // grab the last value
-    } else {
-        if (point <= 0.0) {
-            unit->m_level = level = unit->m_envvals[0];
-        } else {
-            float segpos = point;
-            // determine which segment the current time pointer needs calculated
-            for (int j = 0; point >= newtime; j++) {
-                seglen = unit->m_envvals[(j * 4) + 1];
-                newtime += seglen;
-                segpos -= seglen;
-                stage = j;
-            }
+    unit->m_offset = IN0(1);
+    unit->m_pointin = uninitializedControl; // trigger calculation in next_k
+    unit->m_level = uninitializedControl; // will be set in next_k
 
-            segpos = segpos + seglen;
-            float begLevel = unit->m_envvals[(stage * 4)];
-            int shape = (int)unit->m_envvals[(stage * 4) + 2];
-            int curve = (int)unit->m_envvals[(stage * 4) + 3];
-            float endLevel = unit->m_envvals[(stage * 4) + 4];
-            float pos = (segpos / seglen);
+    IEnvGen_next_k(unit, 1);
 
-            GET_ENV_VAL
-        }
-    }
-    OUT0(0) = level;
+    unit->m_pointin = uninitializedControl; // trigger calculation in next_x
```

Comment:
Maybe we shouldn't reset m_pointin at all? I'm thinking that since the env segments are not read again in _next, we can avoid re-calculating (if input "index" didn't change, we can output m_level without reading again). This way this PR would be purely a refactoring, without even introducing the extra read. I don't see any other interaction between m_pointin and internal state, apart from m_level.

If we decide to not reset m_pointin, I think we should add a comment saying that resetting internal state after init is not necessary for IEnvGen.

What do you think?

Reply:
Sure that's reasonable. I was following the pattern that this collection of init sample fixes follows:
Init state -> calculate an init sample -> reset state so first sample == init sample.

There are exceptions already and I think leaving a comment as you suggested still reinforces the suggested pattern.

Updated!